A Short Hike: Implement New Game

chandler05 opened the pull request:
## What is this fixing or adding?
Adds A Short Hike as a supported Archipelago game.
The randomizer client mod for A Short Hike was developed by @BrandenEK.

## How was this tested?
Extensive private tests by me, an apworld has been available to the public for testing every step of the way and a successful public beta test was performed.

## If this makes graphical changes, please attach screenshots.
![A Short Hike 12_5_2023 3_27_58 PM](https://github.com/ArchipelagoMW/Archipelago/assets/66492208/0cf2e66d-b071-4bfe-aa9d-4628abbd8b0d)
![A Short Hike 12_8_2023 3_44_46 PM](https://github.com/ArchipelagoMW/Archipelago/assets/66492208/661dce86-b5ad-45c4-87ca-8db46abaafe8)

ScipioWright commented on worlds/shorthike/Rules.py:
```diff
@@ -0,0 +1,75 @@
+        lambda state: state.has("Wristwatch", player))
+    add_rule(multiworld.get_location("Sue the Rabbit Shoes Reward", player),
+        lambda state: state.has("Headband", player))
+    add_rule(multiworld.get_location("Return to Shell Kid", player),
+        lambda state: state.has("Shell Necklace", player))
+    add_rule(multiworld.get_location("Return to Shell Kid", player),
+        lambda state: state.has("Seashell", player, 15))
```
```suggestion
    add_rule(multiworld.get_location("Return to Shell Kid", player),
        lambda state: state.has("Shell Necklace", player) and state.has("Seashell", player, 15))
```

ScipioWright commented on worlds/shorthike/docs/en_A Short Hike.md:
```diff
@@ -0,0 +1,25 @@
+## What does randomization do to this game?
+
+All items that can be obtained from chests and NPCs are randomized.
+Additionally, all items on the ground are replaced with chests and randomized.
```
```suggestion
Additionally, all items on the ground are replaced with chests and are randomized.
```

ScipioWright commented on worlds/shorthike/docs/setup_en.md:
```diff
@@ -0,0 +1,33 @@
+## Required Software
+
+- A Short Hike: [Steam](https://store.steampowered.com/app/1055540/A_Short_Hike/)
+    - The Epic Games version of A Short Hike will also work.
```
```suggestion
    - The Epic Games Store version of A Short Hike will also work.
```

thegreenrobby replied:
I run this mod on the DRM free version from itch.io, and it's been working flawlessly. Many people were given a copy of this from the itch.io Racial Justice & Palestinian Aid bundles (from 2020 and 2021, respectively), so that's a nontrivial ownership source. Unless you have intent to not "support" this version, the documentation should probably also acknowledge that?

ScipioWright commented on worlds/shorthike/docs/setup_en.md:
```diff
@@ -0,0 +1,33 @@
+## Connecting
+
+A Short Hike will prompt you with the server details when a new game is started or a previous one is continued.
+Enter in the Server Port, Name, and Password in the popup menu that appears and hit connect.
```
```suggestion
Enter in the Server Port, Name, and Password (optional) in the popup menu that appears and hit connect.
```
Have seen more than a few people confused about passwords in the past, adding (optional) here should make it more clear to some that if you don't have a password, you don't need to put anything there.

ScipioWright reviewed (commented):
Overall, it's a straightforward apworld -- which is a good thing. Very minimal comments, mostly nitpicks. Will try out the mod and do various things to try to break generation later.

qwint commented:
silver feather is marked 'progression' but never logically required (should be switched to 'useful', or factored into req's feathers for access rules)
fishing journal is no longer required (and thus no longer progression), this feels untenable for fishmonger lol
I don't think that many of the progression_skip_balancing items should be labeled as such, stick/toy shovel/seashell make sense as they have duplicates and not all are required, but the rest of the trade items should just be progression

golden feathers in pool option should have a min value when using easy/normal progression (like the min 12 for select goals) for a better UX
ex. letting a player set difficulty to easy and golden feathers to 1 and then expecting them to get every location with only 1 feather seems antithetical to the difficulty setting (especially without explicit documentation) 

using get_feather_state to evaluate min_feathers and then building the lambda from there is much faster in access rules than returning the lambda (which doesn't get evaluated during create_rules)
* this does make setting completion_condition a bit more of a headache, so there might be a decent middle ground, suggest playing around with the new test/benchmark/locations.py (skipping worlds that are not yours for quick retesting)
* an alternative solution that i found to get much faster (but is not usable as-is) is removing the options checks from get_feather_state, so potentially being able to pass in min_feathers and a bucket-state could get very fast as well



did playtests (both solo and with other games) using all goals, testing most settings, doing quests in different orders, etc. and my only gripes are:
* secret island peak should not be 3 golden feathers on Normal Difficulty unless i don't understand some mechanics or there is an alternative route i don't know about (but then imo those should be described somewhere)
* I could not consistently reproduce, but sometimes Avery would not respawn where I would expect them to be (happened most often when winning lighthouse race before losing one, they don't seem to respawn at lighthouse start post again)
* also, not having another save slot could come up as frustrating for some async players, so describing that somewhere in discord pins/docs may be helpful



Some minor gripes that I don't think are dealbreakers:
* are the golden chests and skip cutscenes in the yaml needed when they are qol options in the mod?
* At least the section titles in the webhost docs should start with Capital letters
* create_event is defined and never used, can likely be removed
* "what does another world's item look like in " in the docs is usually for how the 'location' will look ingame, with what the pop-up toast looks like being a secondary note. I'd suggest adding in another note about all items being contained in chests

chandler05 commented:
> silver feather is marked 'progression' but never logically required (should be switched to 'useful', or factored into req's feathers for access rules) 

Fixed.
> fishing journal is no longer required (and thus no longer progression), this feels untenable for fishmonger lol 

It's actually surprisingly easy to catch all of the fish without the journal, just because it only exists as a way to keep track of the fish you've caught in the past. I think it should stay as Useful classification.
> I don't think that many of the progression_skip_balancing items should be labeled as such, stick/toy shovel/seashell make sense as they have duplicates and not all are required, but the rest of the trade items should just be progression

Fixed as well.
> golden feathers in pool option should have a min value when using easy/normal progression (like the min 12 for select goals) for a better UX ex. letting a player set difficulty to easy and golden feathers to 1 and then expecting them to get every location with only 1 feather seems antithetical to the difficulty setting (especially without explicit documentation)

 I agree, but I also think it shouldn't be that big of a deal to let the players have more choice in this situation. I strongly dislike the way I have to force set the feathers to 12 for some of the goals, and I wish that I had the ability to update the options UI based on other options like the goal and difficulty, but I don't. IMO this is fine to leave.
> using get_feather_state to evaluate min_feathers and then building the lambda from there is much faster in access rules than returning the lambda (which doesn't get evaluated during create_rules)
> 
> * this does make setting completion_condition a bit more of a headache, so there might be a decent middle ground, suggest playing around with the new test/benchmark/locations.py (skipping worlds that are not yours for quick retesting)
> * an alternative solution that i found to get much faster (but is not usable as-is) is removing the options checks from get_feather_state, so potentially being able to pass in min_feathers and a bucket-state could get very fast as well

I'll mess around with this to see what I can do.
> did playtests (both solo and with other games) using all goals, testing most settings, doing quests in different orders, etc. and my only gripes are:
> 
> * secret island peak should not be 3 golden feathers on Normal Difficulty unless i don't understand some mechanics or there is an alternative route i don't know about (but then imo those should be described somewhere)

I can give this another test and reevaluate the feather count.
> * I could not consistently reproduce, but sometimes Avery would not respawn where I would expect them to be (happened most often when winning lighthouse race before losing one, they don't seem to respawn at lighthouse start post again)

Avery ends up at different spots depending on an internal clock, but that's just vanilla behavior.
> * also, not having another save slot could come up as frustrating for some async players, so describing that somewhere in discord pins/docs may be helpful

:+1:
> 
> Some minor gripes that I don't think are dealbreakers:
> 
> * are the golden chests and skip cutscenes in the yaml needed when they are qol options in the mod?

Are they not already removed from the yaml Archipelago generates?
> * At least the section titles in the webhost docs should start with Capital letters

Are they not already capitalized?
> * create_event is defined and never used, can likely be removed

:+1:
> * "what does another world's item look like in " in the docs is usually for how the 'location' will look ingame, with what the pop-up toast looks like being a secondary note. I'd suggest adding in another note about all items being contained in chests

:+1:

Thanks for looking through this! I'll do my best to get around to addressing all of these issues.

qwint commented:
re fishing journal: fair, I ended up needing it the couple of times in testing i tried going for all fish, but I didn't feel that strongly either way

re max golden feathers: also fair, but i'd like to see more user-facing info about it then

re qol settings and webdoc capitalization: i don't see them in your options.py / doc files so i think i just forgot to double check and am actually playing on an older version (but did code review on current), oops that's on me lol

qwint commented:
oh, also after trying to make my own start inventory from pool functionality after being inspired by yours:
there is already an option for that in the Options API that pops the item from the itempool and replaces it with and item generated from get_filler_item_name()

it should work with just get_filler_item_name() returning "13 Coins" and adding `start_inventory_from_pool: StartInventoryPool` (importing StartInventoryPool from Options)
(i suggest adding it to the top of your Options object so it gets grouped with the other generic items in the template yaml)

Notably though, this is a different option from start_inventory, but well labeled and gives player control, so I think it's worth it.

chandler05 commented:
Ok, I've addressed all of the issues you raised in the last commit.

qwint reviewed (approved):
New updates cover all my issues,
new access rules benchmark for much faster,
I'm not sure I agree on leaving the normal start inventory's functionality to remove the starting items from pool (init like 64 ish) as well, but I'm not so tied to that to not approve the rest of the game.

ScipioWright commented on worlds/shorthike/__init__.py:
```diff
@@ -0,0 +1,169 @@
+    def __init__(self, multiworld, player):
+        super(ShortHikeWorld, self).__init__(multiworld, player)
+
+    def set_rules(self):
+        create_rules()
```
Invalid and gets overwritten later down, please remove this

ScipioWright commented on worlds/shorthike/__init__.py:
```diff
@@ -0,0 +1,169 @@
+        skipped_items = []
+        additional_junk = 0
+
+        for item, count in self.multiworld.start_inventory[self.player].value.items():
```
```suggestion
        for item, count in self.options.start_inventory.items():
```

ScipioWright commented on worlds/shorthike/__init__.py:
```diff
@@ -0,0 +1,169 @@
+        create_rules(self, location_table)
+
+    def fill_slot_data(self) -> Dict[str, Any]:
+        slot_data: Dict[str, Any] = {}
```
This instance of slot_data is unused, you set slot_data equal to something at line 158, so you can delete this line.

ScipioWright commented on worlds/shorthike/__init__.py:
```diff
@@ -0,0 +1,169 @@
+        options = self.options
+
+        for loc in multiworld.get_filled_locations(player):
+            if loc.item.code == None:
```
Just a minor nitpick, should be either `if not loc.item.code:` or `if loc.item.code is None:`

ScipioWright commented on worlds/shorthike/__init__.py:
```diff
@@ -0,0 +1,169 @@
+    and take in the beautiful world around you.
+    """
+
+    game: str = "A Short Hike"
```
```suggestion
    game = "A Short Hike"
```
Unnecessary to type it here, and pycharm complains about it. Something about overriding a class variable.

ScipioWright commented on worlds/shorthike/__init__.py:
```diff
@@ -0,0 +1,169 @@
+from collections import Counter
+from typing import ClassVar, Dict, Any, Type
+from BaseClasses import Region, Location, Item, ItemClassification, Tutorial
```
```suggestion
from BaseClasses import Region, Location, Item, Tutorial
```
Remove unused import

ScipioWright reviewed (approved):
Not a ton of comments, it's mostly fairly minor. Qwint already did a good job going over everything. I'm approving, but also please address the comments.

qwint reviewed (approved):
Tiny nitpick but __init__ and Rules lost their newline at end of file in the recent updates, but otherwise new code looks good (and i'm kicking myself for not remembering the lambda scope thing)

Berserker66 commented on worlds/shorthike/__init__.py:
```diff
@@ -0,0 +1,155 @@
+
+        menu_region.connect(main_region)
+
+        if self.options.goal == 0:
```
For readibility I'd recommend  using names, rather than numbers. I know it's commented, but might as well use the name then?

Berserker66 commented on worlds/shorthike/__init__.py:
```diff
@@ -0,0 +1,155 @@
+            if loc.item.code is None:
+                continue
+            else:
+                data = {
```
Except for chest angle everything here can be gotten via LocationScout, which is preferred as it reduces the multidata filesize.

Berserker66 replied:
And chest angle seems to be static anyway? Might as well bake it into the mod instead of  storing it on the server and tranferring it?

chandler05 replied:
I went ahead and removed this from the slot data entirely. The apworld should be good to go, although I am still awaiting these changes to be made on the client mod itself. I'll let you know once that change is made.

BrandenEK replied:
@Berserker66 We can remove the item name and the player name from the slot data since those only matter once the player collects the item.
The type is used for each location to display a golden chest instead of a regular one if the item is progression.  This needs to be known right after connection before loading the gameplay scene.  This isn't necessary, but is a cool feature so we can remove that also if it is a dealbreaker.
However, I don't want to hardcode the ap ids into the mod, so can we at least leave the mapping of ap_ids to internal_ids?  The huge benefit of this is that whenever we add new locations or modify the ids, players don't have to update their client, as all that data is stored in only one place, the apworld.
If leaving the { internal_id, ap_id } in the list of locations in slot data is alright and works in the mod, I suppose I can also update blasphemous (Which uses this same method) to remove the extra info in slot data there as well to save even more space.

qwint replied:
should just be able to use NetworkItem.flags instead https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/network%20protocol.md#networkitem
which should be in the scouts as well

chandler05 replied:
This will be moved to the client :+1:

qwint commented on worlds/shorthike/__init__.py:
```diff
@@ -88,25 +87,25 @@ def create_regions(self) -> None:
             # Photo
             self.multiworld.completion_condition[self.player] = lambda state: state.has("Golden Feather", self.player, 12)
-        elif self.options.goal == 2:
+        elif self.options.goal == Goal.option_photo:
```
2 is Races not Photo
but also you should just be able to do `self.options.goal == "photo"` with the new Options API

Berserker66 commented on worlds/shorthike/__init__.py:
```diff
@@ -0,0 +1,136 @@
+    options: ShortHikeOptions
+
+    required_client_version = (0, 4, 4)
+
+    def __init__(self, multiworld, player):
+        super(ShortHikeWorld, self).__init__(multiworld, player)
```
```suggestion
```
This doesn't do anything fyi
